keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,952 @@
+[Pod Index Label for StatefulSets and Indexed Jobs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4017-pod-index-label)
+feature with similar characteristics.
+
+Finally, we would like to have the feature soak without unnecessary delay.
+
+#### Can the label be mutable?
+
+There is a potential risk of leaking pods, if the value of the label is changed.
+For example, assume there is a running Job, which is reconciled by the Job
+controller, and has some pods created.
+Then, if the label is switched to the mirroring Kueue controller (which by
+itself does not manage pods). Then, the pods are leaking and remain running.
+
+In order to avoid the risk of pods leaking between the controllers when changing
+value of the `managed-by` label, we make the label immutable (see also
+[label mutability](#label-mutability)).
+
+However, the question remains if we can make the label mutable when the job is
+stopped, similarly, as we do with the `AllowMutableSchedulingDirectives` flag
+which guards mutability of the Job's pod template labels.
+
+It seems possible, and could be handy in [Story 1](#story-1), but it is also not
+a blocker.
+
+It would also the complicate debuggability of the feature.
+
+We decide to keep the label immutable, at least for [Beta](#beta), we will
+re-evaluate the decision for [GA](#ga).
+
+### Risks and Mitigations
+
+#### Ecosystem fragmentation due to forks
+
+The mechanism to disable the main Job controller opens the door for users to
+substitute it with a fork. This may create more fragmentation in the community
+as users may prefer to use their forked controllers rather than contribute
+upstream.
+
+First, this risk, to some extent, exists even today as admins with access to the
+control plane can disable by passing `--controllers=-job,*` in the manifest for
+`kube-controller-manager` (see more info on the `--controllers` flag
+[here](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)).
+
+Second, we believe that users who had the need to fork the Job controller
+already introduced dedicated Job CRDs for their needs.
+
+#### Two controllers running at the same time on old version
+
+It is possible that one configures MultiKueue (or another project) with an
+older version of k8s which does not support the label yet. In that case
+two controllers might start running and compete with Job status updates at the
+same time.
+
+To mitigate this risk we warn about it in Kueue documentation, to use this label
+only against newer versions of Kubernetes.
+
+Finally, this risk will fade away with time as the new versions of
+Kubernetes support it.
+
+## Design Details
+
+#### API
+
+```golang
+const (
+  ...
+	// LabelManagedBy is used to indicate the controller or entity that manages
+	// an Job.
+	LabelManagedBy = "batch.kubernetes.io/managed-by"
+)
+```
+
+#### Implementation overview
+
+We skip synchronization of the Jobs with the `managed-by` label, if it has any
```

Comment:
Two comments:
1) Where will we skip it? will we configure the watch to completely ignore those jobs (i.e., not receive sync calls at all) or will we add a check in the sync func? The former seems more performant, but the latter allows us to potentially add a condition as suggested below.

2) Does it make sense to add a condition indicating that the job will not be reconciled and is ignored by the job controller? could be useful for debugging and informs the user what is going on.

Reply:
Ad 1. I see possible options:
1. At the informer level using label selector, but it does not reduce the memory usage because of SharedInformers in k8s and would require tricky defaulting of the label, more [here](https://github.com/kubernetes/enhancements/blob/72efa2d9759909b6de1460a810a6a4ddc2e6f45d/keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md#using-label-selectors). Also, doesn’t let us add a log line that the sync is skipped
2. Inside `syncJob` - this is was the initial idea, specifically I was thinking about doing it [here](https://github.com/kubernetes/kubernetes/blob/909faa3a9b33161992af0684c5582353b5be6d9b/pkg/controller/job/job_controller.go#L733)
3. Inside `enqueueSyncJobInternal`, before calling `jm.queue.AddAfter`, this would be slightly more performant, because we don’t go via the queue.

I’m now slightly leaning towards (3.) over (2.) in order not to waste memory for the entries in the queue. 

Ad 2. We are not planning to add the condition, these are the reasons:
1. Its usability seems limited (compared to the label) if the job is immutable (as currently planned), because the timestamps don’t bring extra value
2. Conceptually, we want to give full ownership of the Job object to the other job controller, objects mutated by two controllers could actually make debugging more involving.
3. The MultiKueue controller would have to non-trivially reconcile the Job Status. If it just blindly mirrored the status from the worker cluster (as currently planned), then it would remove the condition. Other controllers would need to be careful not to remove the condition either.
4. This means extra QPS per job, and risks conflicts for the status Update operations.
5. if other (non built-in) controllers were to indicate that they skip with a condition, we would have many conditions

Debugging is important , for this purpose we will log the line saying that the syncing is delegated to another controller (available to admins / OnCalls). The user can list jobs with custom value of the label, for example by `kubectl get jobs -lbatch.kubernetes.io/managed-by -ocustom-columns=NAME:.metadata.name,LABEL:".metadata.labels.batch\.kubernetes\.io/managed-by"`. Then, `kubectl get jobs -oyaml --show-managed-fields` can be useful to figure out what modifies the fields.

Let me know what are your thoughts, I will also follow up with KEP update in case other readers have similar questions.

Reply:
Updated the KEP accordingly PTAL.

Reply:
> I’m now slightly leaning towards (3.) over (2.) in order not to waste memory for the entries in the queue.

Make sense.

> We are not planning to add the condition, these are the reasons:

I am not going to push for a condition because it adds some complexity, so I will leave the decision to you; but the five reasons listed above are not very convincing :) Re1: it is certainly useful for the end user to know why their job is seemingly "stuck". Re 2: we have plenty of objects statuses updated by multiple components within k8s, including pods. Re 3: I don't think it will be "non-trivial", we just need to merge the conditions map on update. Re 4: that is fine, one update over the lifetime of the job is certainly not a problem. Re 5: do we really think we will have that situation? I very much doubt it. I don't think anyone will be implementing a full fledged secondary job controller.

Reply:
@ahg-g thanks for the discussion, I think this is a reasonable idea, and I'm happy to go this path if we conclude this is they best way to address debuggability, despite the drawbacks listed, indirectly also reflecting the complexity you mentioned.  I think it is fair to abandon my point 5. 

So the scenario is that the job is "stuck", and goes looking at the conditions with `kubectl get job -oyaml` or `kubectl describe job` commands. Seeing something like this
```
Type: ManagedByCustomController
Status: True
Reason: CustomManagedByLabel
Message: This Job is managed by a custom controller as it has the managed-by label. 
```
would be more telling than a label, as the users often omit inspection of labels (there are many labels, and they don't know the semantics of them). So, there is indeed a trade-off.

OTOH, maybe the concern could be addressed to some extent by the following:
1. documentation warning that jobs with the custom value of the labels are skipped
2. log line (yeah, just for OnCall or admins)
3. the other controller such as MultiKueue adding the condition that it takes control, or emitting an event for the Job (would not be added if say MultiKueue is not installed on the cluster)
4. extra kubectl get column `MANAGED_BY` listing value of the label
5. extra `kubectl describe` line, say This Job is managed by a custom controller as it has the managed-by label."

WDYT? I would also like to hear from others.

Reply:
Ok, I don't feel like having a strong opinion, just not sure currently about the trade-offs. I will openly wait for more feedback. Updated the KEP with the point to re-evaluate different options before going GA.

Reply:
I'll put another :+1: for a condition. For cases like typos in the label it'll be more convenient for users to figure out what's going on. Also, if you're thinking about operating big clusters, or multiple clusters I don't like the idea of involving cluster admins with debugging every stuck job. The controller is responsible for exposing as much information to the user as possible. The cost vs value here is negligible with our current possibilities (patch on .status.conditions).

Reply:
Thanks for the input!

One more thing which keeps me hesitant is that we have a similar case for `spec.schedulerName`, for which don't add the condition (nor event) when there is a custom value. Similarly, typos may occur, and the consequences are similar, as the pod might get stuck pending. Also similarly, the field is immutable. 

I would yet like to hear what makes the Job case different. 

One potential difference I see is that this is intended for multi-clluster Kueue, so every bit to improve debuggability matters. Let me know if there is another relevant difference point I might be missing.

Reply:
I also realize that the concern (3.) about MultiKueue controller can be mitigated completely by a validation rule inside k8s, that the condition cannot be deleted once added, so the rule would give early feedback to framework developers.

Reply:
It's a bit weird for a controller to add a condition to something saying it is ignoring it because the object opted out of management by the standard controller… I don't think we decorate pods that use a non-default scheduler or endpointslices marked as managed by some other controller

Reply:
Just for the sake of discussion, not really to push to add the condition. I look at it differently: explaining an unexpected job state. I don't think people look closely at endpointslices if at all, and it is not uncommon to have a second scheduler or multiple scheduling profiles that set different scheduler names. In this case though, my feeling is that we will highly highly unlikely that we will have a second implementation of the job controller, in fact this proposal is not primarily motivated by that, so making it clear that the local controller is not reconciling it seems useful to me.

Reply:
The condition is definitely not a blocker for beta, but I'd like to see it explicitly called out before GA, so we have a clear decision for yes/no having one.